### ISSUE TITLE
Bind to an interface

### ISSUE THREAD

**Issue by crozone** [transmission/transmission#338]
Currently transmission only allows binding to a fixed IPv4 or IPv6 address. This is problematic if the IP address we want to bind to changes dynamically based on DHCP.

This issue was first brought up on the original transmission trac some 8 years ago:

https://trac.transmissionbt.com/ticket/2313

Work was done to add a `bind-interface-device` parameter which toggled SO_BINDTODEVICE on kernels that supported it, and use a polling technique on systems that did not (OSX/FreeBSD).

The final patch was submitted 17 months ago and can be found here:

https://trac.transmissionbt.com/attachment/ticket/2313/bind-to-interface-r14706.patch

However I can't see any evidence that it was ever encorporated into any current branch in the git.

**Comment by Guanglier** [comment]
By the way, ipv4 bind is working, but ipv6 + ipv4 bind is'nt working ! Does anyone have the same problem ?

**Comment by dsully** [comment]
I'd like to see this integrated as well.

**Comment by bestouff** [comment]
+1

**Comment by cfpp2p** [comment]
I made a TRAC archive by the Wayback Machine September 12, 2018 https://github.com/transmission/transmission/issues/703#issuecomment-434789322 but yes, unfortunately the patch was not saved by the Wayback Machine.

**Comment by crozone** [comment]
The track is back up, but I've uploaded the raw patch to pastebin just in case it goes down again:

https://pastebin.com/HJQbr1xD

**Comment by miki2o** [comment]
Why this hasn't been implemented yet?

**Comment by ckerr** [comment]
Looks like @luoqi-git made a [PR for this](https://github.com/transmission/transmission/pull/518) back while the project was semi-unmaintained.

**Comment by pballer2oo7** [comment]
Did the PR by @luoqi-git provide support for named interfaces or a list of ip addresses?

**Comment by mtraynor7** [comment]
Can we have an update on this please, has there been any progress?

**Comment by cure1245** [comment]
Does that PR actually cover binding to a named interface? It looks like this simply allows binding to multiple IP addresses, whereas this feature request is more about binding to an interface whose IP may change, such as a VPN.

I haven't looked at the code very hard, so I may be missing something.

**Comment by pballer2oo7** [comment]
@cure1245, that's how I read it too so it looks like named interface code still needs to be written.

**Comment by mtraynor7** [comment]
Is there any likelihood of this being incorporated into the v4 development?

Pretty please, devs 🙂

**Comment by ckerr** [comment]
This is one of those features that will be really useful to <1% of users, and will require extra testing for configurations AFAIK the maintainers don't currently have.

[As I said last February](https://github.com/transmission/transmission/issues/338?notification_referrer_id=MDE4Ok5vdGlmaWNhdGlvblRocmVhZDI0MjI1NzYzNTo3MDM4MQ%3D%3D#issuecomment-1051315309), If anyone wants to update that patch to the current codebase, to test it, and to submit it as a PR, that would be great. And probably more productive use of our time than driving by every once in awhile to ask if I've written it yet :smile_cat:

**Comment by mtraynor7** [comment]
If I had the necessary skills to submit a PR a I would be happy to, but unfortunately I am just a humble user. I would also argue >1% of users would appreciate this feature - though perhaps there may be a number who don't understand it or the risks/benefits associated with it.

In the meantime I will continue to live in hope that somebody with the right skill set finds this enhancement request as desirable as I do 🙂

**Comment by kuwaitbinary** [comment]
According to the [documentation](https://github.com/transmission/transmission/blob/main/docs/Editing-Configuration-Files.md) they have .. What will solve the problem is just writing this line in terminal then restart Transmission

defaults write org.m0k.transmission BindAddressIPv4 "192.168.7.187" <-- represent your ip .. mine is 192.168.7.187

## macOS
### Overview
macOS has a standardized way of saving user preferences files using [XML](https://en.wikipedia.org/wiki/XML) format. These files are called [plist](https://en.wikipedia.org/wiki/Plist) (short for property list) files. Usually there is no need to modify these files directly, since Apple provided a [command-line tool](https://developer.apple.com/DOCUMENTATION/Darwin/Reference/ManPages/man1/defaults.1.html) to reliably change settings. You do need to restart Transmission before these have effect.

In short:
 * To set a key: `defaults write org.m0k.transmission <key> <value>`
 * To reset a key: `defaults delete org.m0k.transmission <key>`
 
 
 
<img width="642" alt="Screenshot 2023-02-24 at 5 50 41 PM" src="https://user-images.githubusercontent.com/1690779/221209625-2ce00039-3cdb-4bca-af52-a08ffc4aa79c.png">

**Comment by cure1245** [comment]
> According to the [documentation](https://github.com/transmission/transmission/blob/main/docs/Editing-Configuration-Files.md) they have .. What will solve the problem is just writing this line in terminal then restart Transmission
> 
> defaults write org.m0k.transmission BindAddressIPv4 "192.168.7.187" <-- represent your ip .. mine is 192.168.7.187

Please read the issue more carefully. The issue is not about binding to IP addresses, the issue is binding to physical interfaces. If the IP address for a given interface is dynamic, there is currently no way to force it to use that particular interface.

**Comment by kuwaitbinary** [comment]
> documentation

If I am expert in Bash I will do a cron for something like this

WARNING: IS NOT TESTED

```
#!/bin/bash
OLD_IP=`cat ip.txt`
NEW_IP=`/sbin/ifconfig  | awk -F "[: ]+'{ print $4}'`
if [ $NEW_IP != OLD_IP ]; then
    pkill transmission
	defaults write org.m0k.transmission BindAddressIPv4 $NEW_IP
	sleep 120
	open -a /Applications/Transmission.app/
    echo $NEW_IP > ip.txt
fi
```

**Comment by benlc** [comment]
I'd love to see that happen

**Comment by adiastra** [comment]
PLEASE allow us to bind to a specific interface, You realize that MANY of us are using VPNs right?

**Comment by benlc** [comment]
By "MANY" you probably mean "MOST"

**Comment by luoqi-git** [comment]
For some reason I've been receiving emails about this ticket, so let me offer some advice. What most people here really want is routing control, to direct the torrent traffic through a particular interface, an uplink with dynamic address or a vpn tunnel. This is best/more flexibly achieved through firewall rules than adding specific support in individual applications. It is also the most portable approach, as there is no standard for bind-to-interface.

Here's roughly how it works:
- assign a unique ip to an internal interface
- bind to this unique ip
- configure the firewall to route all traffic originating from this unique ip to the desired outgoing interface

I'm not going to show you the detailed commands (which by the way depend on the firewall you're using), simply google for "source based routing" and of course the name of your favorite firewall.

**Comment by cure1245** [comment]
> - assign a unique ip to an internal interface

Oh! That just got me thinking: those of us on Linux could probably just create a static alias for the VPN device! Wouldn't even need to mess with the firewall

**Comment by ckerr** [comment]
> PLEASE allow us to bind to a specific interface, You realize that MANY of us are using VPNs right?

> The most missing feature!

> By "MANY" you probably mean "MOST"

https://github.com/transmission/transmission/issues/338#issuecomment-1375960257

**Comment by stunaz** [comment]
@luoqi-git I think what most people wants (surely myself) its to use transmission only via a vpn. qtorrent has it, its like 
![](https://www.downloadprivacy.com/wp-content/uploads/2019/12/interface-binding.png)

**Comment by mispp** [comment]
Also, would be nice to be able to bind API/web access to a different interface from the one used for data transfer

**Comment by securingmom** [comment]
how is qBittorrent, a ui for transmission, able to achive this?  (tun*)

**Comment by RedAnon** [comment]
> how is qBittorrent, a ui for transmission, able to achive this? (tun*)

qBittorrent doesn't use transmission as a backend
qBittorrent uses libtorrent instead

**Comment by 9696neko** [comment]
Would be nice to have this feature as well! I had to pass on Transmission because of the lack of this option.

**Comment by garoto** [comment]
I'm still trying to process @luoqi-git seemingly insightful comments on this issue and the PR itself to this day, but I guess I'm just to dumb to grasp the full content on how binding to a specific nic address can be so elusive.

**Comment by crozone** [comment]
@gonstr Tbh not sure what @luoqi-git is saying with that last comment.

Linux supports `SO_BINDTODEVICE` in `setsockopt` which:

> Bind this socket to a particular device like “eth0”, as specified in the passed
      interface name.  If the name is an empty string or the option length is zero,
      the socket device binding is removed. The passed option is a variable-length
      null-terminated interface name string with the maximum size of IFNAMSIZ.
      If a socket is bound to an interface, only packets received from that particular
      interface are processed by the socket.  Note that this works only for some socket
      types, particularly AF_INET sockets. It is not supported for packet sockets (use
      normal bind(2) there).
      
Sample:

```c
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <net/if.h>

int main(void)
{
    const int sockfd = socket(AF_INET, SOCK_STREAM, 0);

    if (sockfd < 0) {
        perror("socket");
        return EXIT_FAILURE;
    }

    const struct ifreq ifr = {
        .ifr_name = "enp0s3",
    };

    if (setsockopt(sockfd, SOL_SOCKET, SO_BINDTODEVICE, &ifr, sizeof(ifr)) < 0) {
        perror("setsockopt");
        return EXIT_FAILURE;
    }

    const struct sockaddr_in servaddr = {
        .sin_family      = AF_INET,
        .sin_addr.s_addr = inet_addr("142.250.73.196"),
        .sin_port        = htons(80),
    };

    if (connect(sockfd, (const struct sockaddr*) &servaddr, sizeof(servaddr)) < 0) {
        fprintf(stderr, "Connection to the server failed...\n");
        return EXIT_FAILURE;
    }

    // Make an HTTP request to Google
    dprintf(sockfd, "GET / HTTP/1.1\r\n");
    dprintf(sockfd, "HOST: www.google.com\r\n");
    dprintf(sockfd, "\r\n");

    char buffer[16] = {};
    read(sockfd, buffer, sizeof(buffer) - 1);

    printf("Response: '%s'\n", buffer);

    close(sockfd);
    return EXIT_SUCCESS;
}
```

So basically yes, this will act like a packet "filter" and ignore every packet destined for a socket that didn't come in on the specified interface. This is basically what everyone means when they say "bind to an interface".

Given that the additional code is remarkably simple (it's basically just a few lines to do the `setsockopt` with `SO_BINDTODEVICE`), I might actually try and get a PR going for this myself. The hardest part is going to be threading through a config value, but it should basically be the exact same as binding to an IP address.

**Comment by seguri** [comment]
I'd like to add that the default behaviour of `bind-address-ipv4` is very insecure. I issued these commands hoping they would prevent Transmission from connecting outside of Mullvad, and instead the client starts normally if Mullvad is not connected:

```
# Will this disable ipv6?
defaults write org.m0k.transmission bind-address-ipv6 '::1'
defaults write org.m0k.transmission bind-address-ipv4 $(ifconfig $(mullvad status -v | awk -F '[: ]+' '/Tunnel interface/ {print $3}') | awk '/inet / {print $2}')
```

(For brevity ignore that `ifconfig $(mullvad status -v)` is equal to `ifconfig` when VPN is disconnected)

**Comment by tearfur** [comment]
I want to replace `bind-address-ipv*` settings with interface names in `5.0.0`. Hopefully I'll still be around by the time Transmission gets to `5.0.0-dev`.

**Comment by wellloaded** [comment]
Is there any progression on this please?

**Comment by MasterCATZ** [comment]
I have a feeling Transmissions trying to use the wrong network interface on my Linux box ,99% sure its trying to use virtual machines bridge , and would like an option to manually force it to use the correct one as well ...

**Comment by cure1245** [comment]
> Is it possible that this will be implemented for 4.1?

As someone who has been tracking this open feature request for three years _out of the eight it has been open_, I think it's fair to say the maintainer(s) view this as far too much effort for too little benefit (in fact, I believe they've said as much in this very thread).

I'd call myself a fairly quick learner, but after spending a few hours trying to find a good spot to start diving in on my own PR, I gave up. The fact is that Transmission is (lovingly, and with due respect to the maintainers and anyone who has contributed) an old battle tank of an application, and there's bits and pieces of netcode _everywhere_. It probably wouldn't require a total rewrite, but I guarantee anyone who has both the chops and the moxie to start to tackle this simply doesn't have the time to spend on free work.

If _you_ happen to be that kind of person (and can manage a small team of willing would-be contributors [again, for free]), I would be more than willing to assist with the grunt work. Until that person comes along, however, I don't think this will be happening.

**Comment by crozone** [comment]
@cure1245 I attempted a PR myself and came to the same conclusion. Figuring out where to apply the socket options was difficult in itself, but I also could not get my head around how to thread through the configuration options. I aim to make another attempt at a PR eventually but coming into the code fresh, it really is a beast.

**Comment by pa-0** [comment]
@crozone It looks like someone created a separate app already to handle this: [Transmission Interface Binder](https://github.com/neverfox/Transmission-Interface-Binder).  Not sure if it's deprecated now since it apparently has not been maintained, but might be easier to update the code of this utility rather than creating a PR for transmission.

Edit (1): There's also [Clutch](https://github.com/tagyro/Clutch) which is also specifically for Transmission

Edit (2): Found one more: [BindToInterface](https://github.com/JsBergbau/BindToInterface) - This one appears to be for general use in binding both unix and linux apps to specific network interfaces.  It was also published more recently so better chance it still functions properly 😄

**Comment by jameshealyio** [comment]
Forked, PR'd -- made UI changes for macOS GUI, tested and working on CLI + macOS. https://github.com/transmission/transmission/pull/8778/commits

**Comment by cure1245** [comment]
Not the hero we deserve, but the hero we need.

**Comment by jameshealyio** [comment]
If you're not a maintainer and you've compiled and run [my PR](https://github.com/transmission/transmission/issues/338#issuecomment-4332403540) on macOS or CLI; if you have feedback on the implementation, or put any comments on the PR, please do. I'm a little unsure about the UX on the macOS GUI.

**Comment by robd003** [comment]
@jameshealyio it's working great! UI is exactly what I'd expect :^)
